spring-cloud-kubernetes-test-support/src/main/java/org/springframework/cloud/kubernetes/integration/tests/commons/Commons.java

```diff
@@ -168,6 +172,54 @@ public static void loadImage(String image, String tag, String tarName, K3sContai
 
 	}
 
+	/**
+	 * either get the tar from '/tmp/docker/images', or pull the image.
+	 */
+	public static void load(K3sContainer container, String tarName, String imageNameForDownload, String imageVersion) {
+		File dockerImagesRootDir = Paths.get(Commons.TMP_IMAGES).toFile();
+		if (dockerImagesRootDir.exists() && dockerImagesRootDir.isDirectory()) {
+			File[] tars = dockerImagesRootDir.listFiles();
+			if (tars != null && tars.length > 0) {
+				Optional<String> found = Arrays.stream(tars).map(File::getName).filter(tarName::contains).findFirst();
+				if (found.isPresent()) {
+					LOG.info("running in github actions, will load from : " + Commons.TMP_IMAGES);
+					Commons.loadImageFromPath(found.get(), container);
```

Comment:
my 2p is to make the GH workflow do this, as it is distracting in the normal code. In my go code, I only pull if the image doesn't already exist (or it is a locally built one). So, if you can have in the GH action the images safely restored from tar, then "pull if not exists" here for the non-test/tagged images will be simpler code. Also, it centralizes the responsibility vs picking a magic tar directory. When I look at other code that uses k3s or testcontainers it is a lot simpler than the code here, and I think we need to keep in mind how difficult this is vs the intent of testcontainers (just use it sort of thing)

Reply:
yeah in fact when I have something complicated for GH action, I put it into a shell script and have the GH action call that. This way it is easier to ad-hoc test. I have it on my TODO to look at https://dagger.io/ instead of this approach

Reply:
Sorry if I was unclear. What I meant was to just have one branch like you have below, and only focus on docker having the image or not. Agree there is time in ctr you could try to avoid, but I really do think "the cure is worse than the disease" special casing this here. In any case, you could try first a change that only focuses on docker and then measure to see if the technical debt around skipping a step to use ctr directly is worth it or not. For example, pare this PR down only about docker, then measure impact in CI, then you can raise something that does ctr directly and decide.

```
			Commons.pullImage(imageNameForDownload, imageVersion, container); // <- here pull if not present
			Commons.loadImage(imageNameForDownload, imageVersion, tarName, container); // <- here always uses ctr
```

Reply:
IMHO the goal is that the upstream testcontainers library handles logic like this, as it is quite the abstraction break to have to review it in a leaf project like this.

For example, in Go, there's already a function to load images, and while this part is slow (I do in it goroutines, so at least it is not so bad in context), it is better than having to think so hard for something that should be built-in https://github.com/testcontainers/testcontainers-go/blob/main/modules/k3s/k3s.go#L186-L218

Put another way.. I think ideal case is the loadimages code is in testcontainers (java), as this also gets you free advice from other people who deal with this stuff. Right now, it is only you and me (always ;) ) because this project seems nobody else knows or has an opinion on it.

Reply:
TL;DR; so, (k3s) loadImages should be commodity, and saving/restoring (docker) images (safely without fork leak) could also be commodity. Both are different abstractions and should work in their own places vs doing CI logic here.

Reply:
thanks for raising it. Most of the time testcontainers-java is ahead of go, and this is a rare place where we need to catch up the main one. 👍

Reply:
Just to be clear when we run the tests locally (or internally on Jenkins) we will still pull the images like we did before because this cache will never exist?